### PR TITLE
feat(cxl-ui): implement dashboard card design

### PR DESCRIPTION
**Pull request by freudFlintstone**
Task: https://app.clickup.com/t/861n0y1vn

[Design specs](https://www.figma.com/file/RTj2PtUSrCOUFtljOiq0W3/Dashboard?type=design&node-id=171-9732&mode=dev)

### CODE REVIEW

**Comment by freudFlintstone** [comment]
Stories:
Screenshot from 2023-07-07 15-17-15
Screenshot from 2023-07-07 15-16-23

**Comment by anoblet** [comment]
I would add dependencies to the PR description:

Clickup: https://app.clickup.com/t/861n0y1vn
Figma: [Design specs](https://www.figma.com/file/RTj2PtUSrCOUFtljOiq0W3/Dashboard?type=design&node-id=171-9732&mode=dev)

Dependencies:
- https://github.com/conversionxl/aybolit/pull/289

**Review by anoblet — CHANGES_REQUESTED** [review]

**Comment by freudFlintstone** [comment]
> I would add dependencies to the PR description:
> 
> Clickup: https://app.clickup.com/t/861n0y1vn Figma: [Design specs](https://www.figma.com/file/RTj2PtUSrCOUFtljOiq0W3/Dashboard?type=design&node-id=171-9732&mode=dev)
> 
> Dependencies:
> 
>     * [feat: add dashboard lightweight card #289](https://github.com/conversionxl/aybolit/pull/289)

Oh, but I did not build on top of that branch. I actually discussed that with @heshfekry, that we would probably have some overlap, but @kertuilves was too far along with her work then, I think, although not ready for me to build on it. Should I wait for her merge and the rebase my branch?

File: packages/cxl-ui/scss/cxl-card.scss
Comment (freudFlintstone) [reply packages/cxl-ui/scss/cxl-card.scss:3]: Good point. To prevent that, the new styles for the dashboard are scoped under `theme="cxl-course"`, so it doesn't affect the current design instances at all. 

But there's a similar issue in the other direction, meaning the new design is not 100% correct, as some tokens would have to be updated globally, like border-radius. I figured overriding it temporarily is unnececessary.

Comment (anoblet) [reply packages/cxl-ui/scss/cxl-card.scss:3]: While it more be more expedient to introduce it in a feature PR, it affects the visual output of other PR. This change is fairly benign, though in the future we should consider isolating changes.

**Comment by anoblet** [comment]
My mistake. Are we able to add a story for `cxl-card[theme="cxl-dashboard"]`.?

File: packages/cxl-ui/scss/cxl-card.scss
Comment (pawelkmpt) [inline packages/cxl-ui/scss/cxl-card.scss:43]: There might be multiple themes applied at the same time

```suggestion
:host([theme~="cxl-dashboard"]) {
```

File: packages/cxl-ui/src/components/cxl-card.js
Comment (pawelkmpt) [inline packages/cxl-ui/src/components/cxl-card.js:69]:
```suggestion
              alt="${this.instructor}"
```

File: packages/cxl-ui/src/components/cxl-card.js
Comment (pawelkmpt) [inline packages/cxl-ui/src/components/cxl-card.js:23]: I'd rename it to "content tags" or "secondary tags". Body tags is confusing

**Review by pawelkmpt — CHANGES_REQUESTED** [review]

File: packages/cxl-ui/src/components/cxl-card.js
Comment (freudFlintstone) [reply packages/cxl-ui/src/components/cxl-card.js:44]: with current cxl-card usage, like theme=cxl-testimonial theme. I'll improve the comment for future reference.

**Comment by freudFlintstone** [comment]
> My mistake. Are we able to add a story for `cxl-card[theme="cxl-dashboard"]`.?

My mistake, I was actually in the process of changing all of it to `theme=cxl-course`, or `theme=cxl-course-card`. I found it confusing to describe the two stories without separating the container's **dashboard** layout from the card **course** theme.  Your thoughts, @pawelkmpt @anoblet ?

File: packages/storybook/cxl-ui/cxl-card/template.js
Comment (pawelkmpt) [inline packages/storybook/cxl-ui/cxl-card/template.js:35]: Add empty newline

File: packages/cxl-ui/src/components/cxl-card.js
Comment (pawelkmpt) [inline packages/cxl-ui/src/components/cxl-card.js:31]: @freudFlintstone you mentioned that @kertuilves used different approach for this state of the card. This looks good at the first glance but what do you think about introducing `status` attribute and property which would be more universal? It could have `new`, `completed` etc.

**Review by pawelkmpt — CHANGES_REQUESTED** [review]
1. Please clean up git. You need to rebase on top of `feat/dashboard`.

2. Card has "Read more" part. What happens if there's no "more" to show?

File: packages/cxl-ui/src/components/cxl-card.js
Comment (freudFlintstone) [reply packages/cxl-ui/src/components/cxl-card.js:31]: Sure, sounds good!

**Comment by freudFlintstone** [comment]
> 1. Please clean up git. You need to rebase on top of `feat/dashboard`.

I tried that, but it looks like `feat/dashboard` is actually a little behind `master`, on which my branch was based. Can I rebase `feat/dashboard` from `master` myself?

**Comment by freudFlintstone** [comment]
>     2. Card has "Read more" part. What happens if there's no "more" to show?

Not specified in the design. Should I just hide or disable it? What do you think, @heshfekry ?

**Comment by pawelkmpt** [comment]
> > 1. Please clean up git. You need to rebase on top of `feat/dashboard`.
> 
> @pawelkmpt I tried that, but it looks like `feat/dashboard` is actually a little behind `master`, on which my branch was based. Can I rebase `feat/dashboard` from `master` myself?

You can

File: packages/cxl-ui/scss/cxl-course-card.scss
Comment (pawelkmpt) [inline packages/cxl-ui/scss/cxl-course-card.scss:40]: I think these should be default styles for a card which is named `cxl-course-card`. At the same time we need other color variations and for them I suggest to create themes:
* video,
* playbook,
* training.

File: packages/cxl-ui/scss/cxl-course-card.scss
Comment (pawelkmpt) [inline packages/cxl-ui/scss/cxl-course-card.scss:124]: Font size in this section should be the same as in other parts

File: packages/cxl-ui/src/components/cxl-course-card.js
Comment (pawelkmpt) [inline packages/cxl-ui/src/components/cxl-course-card.js:26]: I think `avatar` is sufficient

File: packages/cxl-ui/src/components/cxl-course-card.js
Comment (pawelkmpt) [inline packages/cxl-ui/src/components/cxl-course-card.js:45]: It is new component now. Is it still needed as before?

**Review by pawelkmpt — CHANGES_REQUESTED** [review]
Looking good 👍 

1. Storybook: we need preview for all kinds of cards (video, playbook, etc.)
2. Does your editor respect `.editorconfig` rules? We have `insert_final_newline = true` there but you keep pushing files without final newline.

**Review by pawelkmpt — CHANGES_REQUESTED** [review]
Adjust commit message

```diff
- feat(cxl-ui): implement dashboard card design
- feat(cxl-ui): add `cxl-course-card` component
```

**Review by pawelkmpt — COMMENTED** [review]
I just made a review @freudFlintstone and don't see all changes applied

File: packages/cxl-ui/scss/cxl-course-card.scss
Comment (freudFlintstone) [reply packages/cxl-ui/scss/cxl-course-card.scss:40]: I completely missed the different background in figma. My bad. However, we already have to set `type="Course|Video|Training|Playbook"`. That attribute is used both to determine the type in logic and as a text value for the first tag at the top of the header. Can I just use that to set the CSS background?

**Comment by freudFlintstone** [comment]
@pawelkmpt I"ll check what's going on with the final line, sorry about that.

File: packages/cxl-ui/scss/cxl-course-card.scss
Comment (freudFlintstone) [reply packages/cxl-ui/scss/cxl-course-card.scss:124]: In the footer section, font-size is only adjusted for **::part(summary)** and **::part(toggle)** of the vaadin-details components header, to match the figma design as close as possible. Text content is not affected.

File: packages/cxl-ui/src/components/cxl-course-card.js
Comment (freudFlintstone) [reply packages/cxl-ui/src/components/cxl-course-card.js:45]: No, you're right, taking it out

**Comment by freudFlintstone** [comment]
@pawelkmpt Some observations for this review:
- added stories for all card types
- background colors are not available from cxl-lumo-styles, so the colors are set in the this component scope only until we assess what tokens might be needed, considering other parts of the redesign
- Card heights might need a design decision. For now, I chose to follow the figma specs and set a `min-height: 270px`, but the downside is that opening the "read more" section will use flexible space above it, if description is short, which then makes the read more + CTA button move up and down when toggling that section.

**Comment by pawelkmpt** [comment]
> * added stories for all card types

Thank you! Look good 👍 

> * background colors are not available from cxl-lumo-styles, so the colors are set in the this component scope only until we assess what tokens might be needed, considering other parts of the redesign

👍 

> * Card heights might need a design decision. For now, I chose to follow the figma specs and set a `min-height: 270px`

It's good you chose what Figma defined

> but the downside is that opening the "read more" section will use flexible space above it, if description is short, which then makes the read more + CTA button move up and down when toggling that section.

I noticed this behavior. Can we fix it? It can't work like that

**Review by pawelkmpt — CHANGES_REQUESTED** [review]

File: packages/cxl-ui/scss/cxl-course-card.scss
Comment (freudFlintstone) [reply packages/cxl-ui/scss/cxl-course-card.scss:19]: The first comment is wrong, should be 3pct. The 2pct variant does already exist. Those are there to point that the base colors (primary and shade) could have those variants, but they don't yet exist. The third one, for the video card background, has no similar color anywhere in lumo or our theme.

**Comment by freudFlintstone** [comment]
> > * Card heights might need a design decision. For now, I chose to follow the figma specs and set a `min-height: 270px`
> 
> It's good you chose what Figma defined
> 
> > but the downside is that opening the "read more" section will use flexible space above it, if description is short, which then makes the read more + CTA button move up and down when toggling that section.
> 
> I noticed this behavior. Can we fix it? It can't work like that

So, yes, but not without different cons. @heshfekry, your input would be useful here too. Ideas:

1. Set the card height and the body content height: This would solve it, all cards would look the same height and have the footer in the same position, and also expand when needed without moving the footer. However, what should we do if the body content (course description) is longer than expected? 

2.  Go back to what I saw before this, having only a minimal height. THis means that cards might be taller if the content is longer, but it solves the footer situation. The other glaring downside is that the grid won't look as nice, unless we find a minimal-height that would be enough for most descriptions

**Comment by heshfekry** [comment]
> > > * Card heights might need a design decision. For now, I chose to follow the figma specs and set a `min-height: 270px`
> > 
> > 
> > It's good you chose what Figma defined
> > > but the downside is that opening the "read more" section will use flexible space above it, if description is short, which then makes the read more + CTA button move up and down when toggling that section.
> > 
> > 
> > I noticed this behavior. Can we fix it? It can't work like that
> 
> So, yes, but not without different cons. @heshfekry, your input would be useful here too. Ideas:
> 
> 1. Set the card height and the body content height: This would solve it, all cards would look the same height and have the footer in the same position, and also expand when needed without moving the footer. However, what should we do if the body content (course description) is longer than expected?
> 2. Go back to what I saw before this, having only a minimal height. THis means that cards might be taller if the content is longer, but it solves the footer situation. The other glaring downside is that the grid won't look as nice, unless we find a minimal-height that would be enough for most descriptions
> 
> If it helps, I can make both these cases a story

Option 1 has the most legs because option 2 destroys the grid. The View course CTA moving up and down is less of a problem. With content longer than normal, lets see an example and assess. Logic solution would be to increase the overall size of the card a little to make space for a specific number of characters, and we use that to cap the content of the description in course production. This should have a limit anyways. What is the character limit now?



On another note - I see dodgy spacing and size on bottom left card here.

image

**Comment by freudFlintstone** [comment]
I'll also make sure those tags don's have that odd spacing when there are many of them.

>     3. How is the new state applied?
just a boolean attribute on the card:
```
<cxl-course-card new>
 <content>
</cxl-course-card>
```

**Comment by freudFlintstone** [comment]
@heshfekry After alot of testing, I feel like 300px of height is a good value. It will handle 4 lines of description easily. Even if tags and title wrap, it's doesn't stretch that  much. I added another dashboard story to show some extreme examples.

**Review by pawelkmpt — CHANGES_REQUESTED** [review]
> @heshfekry @pawelkmpt After a lot of testing, I feel like 300px of height is a good value. It will handle 4 lines of description easily. Even if tags and title wrap, it's doesn't stretch that much. I added another dashboard story to show some extreme examples.

@freudFlintstone it keeps jumping. I experimented with `min-height` for `<p slot="content"></p>` element in and `min-height: 110px;` solves the problem. Is there anything which stops us from using such value?

Also - what I already mentioned at least once: **`more` slot needs smaller font**. The same size as other parts of the card

**Comment by freudFlintstone** [comment]
> > @heshfekry @pawelkmpt After a lot of testing, I feel like 300px of height is a good value. It will handle 4 lines of description easily. Even if tags and title wrap, it's doesn't stretch that much. I added another dashboard story to show some extreme examples.
> 
> @freudFlintstone it keeps jumping. I experimented with `min-height` for `<p slot="content"></p>` element in and `min-height: 110px;` solves the problem. Is there anything which stops us from using such value?

No, I just haven't tried using that because it's a big difference from the design spec and short descriptions will have a lot of white-space below. But if having a square grid is important, then setting exact heights is a requirement. Also, itś not a fix for everything. As you can see below, we need a fixed height for the title too, which risks an even larger title being cut-off. I'll do some testing on that.
image


> Also - what I already mentioned at least once: **`more` slot needs smaller font**. The same size as other parts of the card
Fixing it now

**Review by pawelkmpt — CHANGES_REQUESTED** [review]
https://www.loom.com/share/2e01dc825c5c4a76a4a9e28461e2817a?sid=f41c2c43-125a-4468-b3b8-76658ffd14f0

    @pawelkmpt I'm re-requesting the review, but we still need to [make a decision](https://cxlworld.slack.com/archives/C01HXUNGEPM/p1689686050275289) about whether to enforce some layout constraints or just let it go for now.

TLDR; min-height constraints made cards look worse the before. Let’s revert.

Main thing I didn’t like about cards behavior was:

    opening the "read more" section will use flexible space above it, if description is short, which then makes the read more + CTA button move up and down when toggling that section.

As I understand, in order to make “read more” part expand an the bottom and make card height bigger without “up and down jump” effect is setting min-heigh for all card elements (title, content, …)?

If so, we should just accept how it works for now and avoid spending days on it.

Title:

    no min height -> too much whitespace below for short titles;
    no line limits -> user needs to know the full name of the course;

Title tags: just one line like on the design.

Content tags:

    do not limit them to one line. They should wrap with small line-height like shown below;
    line clamping -> as mentioned in Loom video - we can’t have multiple open/closed states: on hover and on click. It’s too much for such simple component. Remove it.

254796484-4c5adc5c-996a-4a24-b135-dec1571d9cd1

Important: before overwriting code, make a backup just in case we'd like to revisit these ideas again.

**Comment by freudFlintstone** [comment]
No problem, I'll revert all changes manually.

> Important: before overwriting code, make a backup just in case we'd like to revisit these ideas again.

I'd have to disagree on this one. The right way to do it is to leave the real commits, so we can revert or cherry-pick using the proper git tools if needed. We can always squash before merging. My apologies if that's already what I was supposed to do, but I understood I should always squash, even during PR process.

File: packages/cxl-ui/scss/cxl-course-card.scss
Comment (pawelkmpt) [inline packages/cxl-ui/scss/cxl-course-card.scss:166]: I see other component use `var(--lumo-line-height-*)` values. I experimented with `line-height` here and `--lumo-line-height-s` is the closest and I recommend changing values to lumo vars to keep consitency.

**Review by pawelkmpt — CHANGES_REQUESTED** [review]
@freudFlintstone I noticed component looks bad on small mobile 320px but even 360. Everything should look good in 320px. Please correct it.

Screenshot 2023-07-24 at 09 29 10

Tags are cut in the bottom

Screenshot 2023-07-24 at 09 30 44

File: packages/cxl-ui/scss/cxl-course-card.scss
Comment (freudFlintstone) [reply packages/cxl-ui/scss/cxl-course-card.scss:166]: Good catch, @pawelkmpt, I did not see the tags being cut on the second line there.  

About the 320px. example, it looks like the header tags are the main offender. Can I just prevent them from wrapping and hide the overflow with an ellipsis, and let the user hover to see a default tooltip (standard HTML title attribute) with the full list of tags? A little less padding under 360px might also help a lot

Comment (pawelkmpt) [reply packages/cxl-ui/scss/cxl-course-card.scss]: > About the 320px. example, it looks like the header tags are the main offender. Can I just prevent them from wrapping and hide the overflow with an ellipsis, and let the user hover to see a default tooltip (standard HTML title attribute) with the full list of tags?

Sure thing

**Review by pawelkmpt — CHANGES_REQUESTED** [review]
It's almost perfect. I was about to merge and then noticed bottom whitespace below CTAs is bigger then before... Please fix it.

Screenshot 2023-07-25 at 09 03 18

**Review by pawelkmpt — APPROVED** [review]

**Pull request merged**
